### ISSUE TITLE
Problem with dash::memalloc 

### ISSUE THREAD

**Issue by devreal** [dash-project/dash#342]
I am using `dash::memalloc` in my `ThreadSafetyTest`s. I'm pretty sure it had worked before but now every call returns a global pointer with offset 0: `dash::GlobPtr<i>(000000|00|0000|0000|0000000000000000)`. 

I have created a simple  test case that currently fails: https://github.com/dash-project/dash/blob/feat-dart-reducempicalls/dash/test/memory/LocalAllocatorTest.cc#L19

Output:
```
[  ERROR   ] [UNIT 0] in /home/joseph/src/dash/dash/dash/test/memory/LocalAllocatorTest.cc:28
Expected: (ptr1) != (ptr2), actual: dash::GlobPtr<i>(u000000|f00|s0000|t0000|o0000000000000000) vs dash::GlobPtr<i>(u000000|f00|s0000|t0000|o0000000000000000)
```

Two allocations should never get the same pointer if there has not been a free in between. 

The same test case with DART functionality succeeds: https://github.com/dash-project/dash/blob/feat-dart-reducempicalls/dash/test/dart/DARTMemAllocTest.cc#L40

That test succeeds. Could this be related to the memory spaces that have been added recently?

**Comment by fuchsto** [comment]
Pretty sure that's on me. `dash::memalloc` used to create a "free-floating" memory space with rather arbitrary global pointer iteration. The type dependencies are fine now, but apparently I broke something else there.
Unit tests passed on my machines with some different MPI versions and compilers, though. I will check your test case now.

**Comment by rkowalewski** [comment]
I am looking into  this.

**Comment by fuchsto** [comment]
I might have tracked it down. @devreal can you please re-build with logging enabled and attach the log snippet around the `dash::memalloc` call?

**Comment by rkowalewski** [comment]
After inspecting what's going on I can see that there is a [TODO](https://github.com/dash-project/dash/blob/feat-dart-reducempicalls/dash/include/dash/GlobPtr.h#L137) by @fuchsto. We do not bind the GlobUnitMem temporary to a global pointer.

So please fix this. I am wondering why these refactorings of the global memory concept are actually at least partially included in this release.

**Comment by fuchsto** [comment]
It's not related to global memory concepts, these are preparations for fixes of global pointer arithmetics.
The `GlobalUnitMem` is not required to be bound as the allocated handle is the `dart_gptr_t` returned by `dart_memalloc` which is later released via `dash::memfree`. If the `GlobUnitMem` would be bound, it would also be deleted in the `GlobPtr`s destructor and that's not what we want. The lifespan of the allocated global memory should only end with `dash::memfree`.
But it's related to the "actual" TODO.

**Comment by rkowalewski** [comment]
Since this temporary is not bound it is destroyed and the destructor actually frees the originally allocated dart global pointer. So both allocations in the mentioned unit tests return a global null pointer.

**Comment by fuchsto** [comment]
Aaah, darn, that's exactly right. I added an ownership flag in `GlobPtr` that is missing in branch development for arcane reasons, seems to have been overwritten in a merge.
It still doesn't quite add up. A global null pointer is not `0000|0000|...`, it's `ffff|0000|....` (undefined unit id is -1). I should have an un-merge-broken version of this in a local workspace.

**Comment by devreal** [comment]
Thanks for investigating. I should note that the pointer is not a `DART_GPTR_NULL`. When you fix this issue, please set the internal `gptr` to `DART_GPTR_NULL`. 

So I guess a log is not needed anymore?

**Comment by devreal** [comment]
Ahh, @fuchsto you were faster. Note that `dart_memfree` does not set the `gptr` to NULL (just like `free()`) so it has to be done manually.

**Comment by fuchsto** [comment]
Dito ;)

I actually set the `dart_gptr_t` to `DART_GPTR_NULL`, the nulliness of the pointer is the actual result of the allocation. But as the memory space is reclaimed before the second allocation, you get the zeroed-pointer value twice. I believe.

**Comment by fuchsto** [comment]
@devreal ... and yes, no log needed!

**Comment by fuchsto** [comment]
Fixed in #344, added the unit test linked in the description, passes on my machine.
